Revised Subgraph and MaskSubgraph

d-michail opened the pull request:
This is a first pass for the revision of subgraph(s) and mask-subgraph(s).

Main changes:
- Revised subgraph and subclasses to use Java 8 features.
- Optimized the initialization of subgraph to also take care of issue #120.
- Revised `MaskSubgraph` by using Java 8 predicates instead of the `MaskFunctor` interface. `MaskSubgraph` was already avoiding copying. Deprecated the `MaskFunctor` interface.

This is relevant to issue #296.

For now I avoided more aggressive optimizations in `SubGraph` like indexing the existing edges. If we want to do that, I think it is best to do this in a second round.

jkinable commented:
Working on it.

On Thu, Oct 13, 2016 at 6:09 PM, John Sichi <notifications@github.com>
wrote:

> *@jsichi* approved this pull request.
>
> Looks good to me. This change could break some apps due to the change from
> copy to virtualization, but I think we can live with that (we may want to
> mention this in the history and release notes).
>
> @jkinable <https://github.com/jkinable> I know you're busy at the moment;
> if I don't see comments from you within the next few days, I'll go ahead
> and merge and if you have comments later, we can address them in a followup.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jgrapht/jgrapht/pull/307#pullrequestreview-4175412>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACnYynJARZ7-Wooza8X8OC8qz2cwNrWdks5qzqwFgaJpZM4KLuNq>
> .
>

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/graph/Subgraph.java:
```diff
@@ -317,16 +304,7 @@ public boolean containsVertex(V v)
-        }
-
-        return edges;
+        return new FilteredSet<>(base.edgesOf(vertex), e -> edgeSet.contains(e));
```
This can be optimized when the Subgraph has an induced edge set.

d-michail replied:
Same here, the `EulerianCircuitTest` never terminated.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/graph/Subgraph.java:
```diff
@@ -207,7 +210,6 @@ public E addEdge(V sourceVertex, V targetVertex)
         for (E e : edges) {
```
I can't seem to put this comment in the right place... you can only comment on modified code. Sorry about that.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/BlockCutpointGraph.java:
```diff
@@ -168,7 +168,8 @@ private void biconnectedComponentFinished(V s, V n)
             new VertexComponentForbiddenFunction(vertexComponent);
-        UndirectedGraph<V, E> biconnectedSubgraph = new UndirectedMaskSubgraph<>(this.graph, mask);
+        UndirectedGraph<V, E> biconnectedSubgraph = new UndirectedMaskSubgraph<>(
+            this.graph, v -> mask.isVertexMasked(v), e -> mask.isEdgeMasked(e));
```
Removed the mask completely.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/graph/DirectedSubgraph.java:
```diff
@@ -51,79 +51,65 @@ public DirectedSubgraph(DirectedGraph<V, E> base, Set<V> vertexSubset, Set<E> ed
-        }
-
-        return edges;
+        return new FilteredSet<>(base.incomingEdgesOf(vertex), e -> edgeSet.contains(e));
```
I somehow wanted to make the behavior more "live", i.e. to reflect the changes that may happen in the base graph.

d-michail commented:
@jkinable After both @jsichi's and your comments, I took back the attempt to provide a more "live" view of the subgraph. It still had some flaws, for example in the case that someone added a vertex in the graph, some of the filtered sets did not reflect that.

Regarding using the subgraph in order to implement algorithms which need a graph copy, I am not sure whether it is easy enough to provide an all use-case solution here. Some algorithms need only a copy of outgoing edges, others both incoming and outgoing, etc. Moreover, the subgraph is designed to reflect the changes that happen in the base graph.

I was thinking that maybe we can provide an implementation of subgraph which indexes the edges that belong to each vertex. This would solve the efficiency problems, but due to the "live" nature of subgraph it is difficult to ensure the same edge iteration order between the subgraph and the base graph (especially if the user keeps adding and removing edges from the base graph).

I have also noticed there are some classes like `DirectedNeighborIndex` and `NeighborIndex' which somehow do the same thing. Maybe we can adjust those, or use those in the implementation of algorithms which need a copy.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/graph/Subgraph.java:
```diff
@@ -142,27 +143,29 @@ public Subgraph(G base, Set<V> vertexSubset)
-                    edges.add(e);
-                }
-            }
+            return new FilteredSet<>(
```
I tried that and the `EulerianCircuitTest` failed (never terminated).

d-michail commented:
I also squashes here the commits, sorry! 

Description of the last changes: 
 - Removed the FilteredSet and used as you suggested the stream collect
 - Removed the class which contains the vertex and edge mask in `BlockCutpointGraph`

Tried but reverted your suggestion regarding the induced optimizations as the `EulerianCircuitTest` did not terminate.

d-michail commented:
A few minor changes to also take care of issue #164.

@jsichi and @jkinable: The current state here is that this pull-request does not change the behavior of the library. The changes are mostly refactoring using Java 8 lambdas and a few optimizations to close 2 issues.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/graph/Subgraph.java:
```diff
@@ -459,6 +419,49 @@ public void setEdgeWeight(E e, double weight)
+        }
+
+        // add edges
+        if (edgeFilter == null) {
```
Can you add one more case:
if( vertexFilter == null && edgeFilter==null) //Subgraph is used as a copy
  edgeSet.addAll(base.edgeSet);

d-michail commented:
Did the changes and added a small paragraph in the javadocs to warn for the possible performance overhead.

jkinable commented:
Great, this looks so much cleaner! very nice work.
